Classes/Emogrifier.php

```diff
@@ -323,6 +334,94 @@ protected function process(\DOMDocument $xmlDocument)
+            // Strip !important indicator
+            $styleValue = trim(str_replace('!important', '', $styleValue));
+            $this->mapCssToHtmlAttribute($styleName, $styleValue, $node);
+        }
+    }
+
+    /**
+     * Applies $styles to $node.
+     *
+     * This method maps a CSS rule to HTML attributes and adds those to the
+     * node.
+     *
+     * @param string $styleName   the name of the style rule to map.
+     * @param string $styleValue  the value of the style rule to map.
+     * @param \DOMNode $node      node to apply styles to.
+     *
+     * @return void
+     */
+    private function mapCssToHtmlAttribute($styleName, $styleValue, \DOMNode $node)
+    {
+        if ($styleName === 'background-color') {
+            // Always override HTML attributes. CSS takes precedence.
+            $node->setAttribute('bgcolor', $styleValue);
+        } elseif ($styleName === 'background') {
+            // Parse out the color, if any
+            $styles = explode(' ', $styleValue);
+            $first = $styles[0];
+            if (!is_numeric(substr($first, 0, 1)) && substr($first, 0, 3) !== 'url') {
+                // This is not a position or image, assume it's a color
+                $node->setAttribute('bgcolor', $first);
+            }
+        } elseif ($styleName === 'width') {
+            $width = preg_replace('/[^0-9.%]/', '', $styleValue);
+            $node->setAttribute('width', $width);
+        } elseif ($styleName === 'height') {
+            $height = preg_replace('/[^0-9.%]/', '', $styleValue);
+            $node->setAttribute('height', $height);
+        } elseif ($styleName === 'margin' && ($node->nodeName === 'table' || $node->nodeName === 'img')) {
+            $margins = preg_split('/\s+/', $styleValue);
+            if (count($margins) === 1) {
+                if ($margins[0] == 'auto') {
+                    $node->setAttribute('align', 'center');
+                }
+            } elseif (count($margins) === 2 || count($margins) === 3) {
+                if ($margins[1] == 'auto') {
+                    $node->setAttribute('align', 'center');
+                }
+            } else {
+                if ($margins[1] === 'auto' && $margins[3] === 'auto') {
+                    $node->setAttribute('align', 'center');
+                }
+            }
+        } elseif ($styleName === 'text-align' &&
+                    ($node->nodeName === 'p' || $node->nodeName === 'td' || $node->nodeName === 'div')) {
+            if ($styleValue === 'left' || $styleValue === 'right' ||
+                    $styleValue === 'center' || $styleValue === 'justify') {
+                // Only set values that are valid for the align attribute
+                $node->setAttribute('align', $styleValue);
+            }
+        } elseif ($styleName === 'float' && ($node->nodeName === 'table' || $node->nodeName === 'img')) {
+            if ($styleValue === 'left' || $styleValue === 'right') {
```

Comment:
I think with a lookup table you don't need a switch or if/elseif. You just go through the levels of the lookup table. But I'm concerned that this function doesn't like to be in a lookup table.

For example, for most properties, you can apply the value to the attribute, but not for all. For margin and background, you need to parse the value of the property. For width and height too, because the 'px' must be stripped. I don't think there's any sane lookup mechanism to map the parse rules (or perhaps a regexp?).

For those properties whose value can be mapped, I think a lookup table is fine. As you suggested, a 2-dim array can do the trick, with one level for the property and a second for a whitelist of tag names.

I think this could work if the leafs of the lookup table can be both strings (the attribute name) and functions (a function to parse the property value and return an attribute-value pair).

Then the simple ones are captured with a simple string and the others with a function.

I'm not convinced this solution will be 'simpler' per se. But perhaps it will be easier to extend without getting lengthy.